### ISSUE TITLE
Make type class instance names optional by generating their names instead

### ISSUE THREAD

**Issue by JordanMartinez** [purescript/purescript#4084]
## Summary

No longer require developers to write a name for type class instances as the compiler will become smart enough to generate a unique name instead.

## Motivation

This change would remove the boilerplate at the source code level which is tedious to write and not necessary. In the real world, one often names such instances using this convention: `className <> fold typeParamterNameList`. For example...
```purescript
class ClassName a b c d
instance classNameFooBarBazBop :: ClassName Foo Bar Baz Bop
```

Rather than forcing the developer to write these names, the compiler would generate them.

## Proposal

Instance names at the source code level would be optional, so that current code doesn't break and future code doesn't need to use additional boilerplate. Perhaps in a future breaking change, such names could be removed altogether.

This change would not remove the requirement that instances have names at the backend-code level (e.g. the outputted JavaScript, etc.). Rather, the compiler would generate a unique name for each type class instance based on the type class name and the type parameters for that class.

Such a generator would need to account for tricky cases while keeping names short yet readable.

## Examples

Rather than writing this:
```purescript
instance classFooBar :: Class Foo Bar where
  func = Tuple Foo Bar

derive instance genericFoo :: Generic Foo _

instance class :: Class { a :: { b :: { c :: Int } }, d :: Tuple Int (Maybe String) }
```
developers would write this:
```purescript
instance Class Foo Bar where
  func = Tuple Foo Bar
  
derive instance Generic Foo _

instance Class { a :: { b :: { c :: Int } }, d :: Tuple Int (Maybe String) }
```
which would be equivalent to writing something like this:
```purescript

instance classFooBar :: Class Foo Bar where
  func = Tuple Foo Bar
  
derive instance genericFoo_ :: Generic Foo _

instance classRecARecBRecCIntDTupleIntMaybeString :: Class { a :: { b :: { c :: Int } }, d :: Tuple Int (Maybe String) }

**Comment by wclr** [comment]
What about deprecating those names and disallowing the usage of instance names in FFI (in future versions)? Just to make codebases more readable in general. I wonder if there really people who would like to keep this meaningless noise in the code.

**Comment by hdgarrood** [comment]
:+1: see also #3426

**Comment by JordanMartinez** [comment]
Also relates to #752

**Comment by JordanMartinez** [comment]
Closed by #4086
